docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+  (or after whatever period is defined in the `--remote-conditions-grace-period` flag)
+- `HealthCheckSucceeded` and `OwnerRemediated` (or `ExternalRemediationRequestAvailable`) are set by the MachineHealthCheck controller in case a resource instance targets the machine.
+- KubeadmControlPlane also adds additional conditions to Machines, but those conditions are not included in the table above
+  for sake of simplicity (however they are documented in the KubeadmControlPlane paragraph).
+
+TODO: think carefully at remote conditions becoming unknown, this could block a few operations ... 
+
+#### Machine Spec
+
+Machine's spec is going to be improved to allow 3rd party to extend the semantic of the new Machine's `Ready` condition
+as well to standardize the concept of Machine's `Availability`.
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```go
+type MachineSpec struct {
+    
+    // MinReadySeconds is the minimum number of seconds for which a Node for a newly created machine should be ready before considering the replica available.
+    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
+    // +optional
+    MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
+
+    // If specified, all readiness gates will be evaluated for Machine readiness.
+    // A Machine is ready when `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are "True"; 
+    // if other conditions are defined this field, those conditions should be "True" as well for the Machine to be ready.
+    // +optional
+    // +listType=map
+    // +listMapKey=conditionType
+    ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`
+
+    // Other fields...
+}
+
+// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
+type MachineReadinessGate struct {
+    // ConditionType refers to a condition in the Machine's condition list with matching type.
+    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
+    ConditionType string `json:"conditionType"`
+}
+```
+
+| v1beta1 (current)      | v1Beta2 (tentative Q1 2025) | v1beta1 removal (tentative Q1 2026) |
+|------------------------|-----------------------------|-------------------------------------|
+| `ReadinessGates` (new) | `ReadinessGates`            | `ReadinessGates`                    |
+| other fields...        | other fields...             | other fields...                     |
+
+Notes:
+- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing this should not trigger rollouts).
+- Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity; 
```

Comment:
Any specific reason for this? I think this might result in unnatural choices for consumers at the time of defining their readiness gates and seems a bit miss aligned with the spirit of the proposal about relaxing custom impositions and standardizing on kube. As stated in kube recommendations It's ok for negative polarity to represent normal operation

Reply:
How would we validate that's only positive polarity?

Reply:
> Any specific reason for this? 

Mostly for sake of keeping this proposal scope under control + I don't have concrete examples that make me think we now need something more complex of what exists in K8s

But I agree with the fact that this is one of the idiosyncrasies in K8s, and so I also added in the next sentence "The Cluster API project might revisit this in the future to stay aligned with Kubernetes or if there are use cases justifying this change."

I also have a few ideas about how to improve this, but if it is not a blocker, I think we should have a minimal feature of this feature in ASAP, and then iterate based on needs.

> How would we validate that's only positive polarity?

We do not validate, it is just what we will assume when computing ready

Reply:
I'm fine with only supporting aggregating positive polarity conditions into Ready for now. It should be relatively trivial to extend the readinessGates field later to also support negative polarity

Reply:
readinessGates should be a slice of some small struct if we see a want to expand it in the future to indicate a desired happy state polarity

Reply:
Oh yup :). That's why we propose
```go
// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
type MachineReadinessGate struct {
    // ConditionType refers to a condition in the Machine's condition list with matching type.
    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
    ConditionType string `json:"conditionType"`
}
```
(which is what is k/k is doing for readinessGates as well)

Reply:
@zjs I assume you mean `Status ConditionStatus`? (same field name as in the condition type)

Reply:
Personally I would prefer to get a first version of this feature implemented without support for negative polarity and then iterate (the API is easily extensible for this use case). Also others agree above.

But I don't have strong opinions, so I would wait for some more feedback. Please, add 👍 if you want to add support for negative polarity now, 👎 if it is ok for you to aim for a first iteration without supporting negative polarity and then follow-up.

In case we implement support for negative polarity, let's keep in mind https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1692281748

Reply:
> aim for a first iteration without supporting negative polarity and then follow-up.

@fabriziopandini Probably good time to start an umbrella issue (where we can also have a section with backlog / follow-ups)

Reply:
Tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105, also the proposal already has note about the intent to consider implementing support for negative polarity conditions to ReadinessGates and AvailabilityGates